### PR TITLE
Adding 'end survey' action to SurveyRule

### PR DESCRIPTION
**Pull request by alxdarksage**
Added a new field indicating the survey should come to an end, a rule either has a skipToTarget or endSurvey = true.

Improved SurveyRule - made it immutable and final, checked with test, cleaned up hashCode, equals() methods.

### CODE REVIEW

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java [hunk 1]
```diff
@@ -15,75 +18,64 @@
...
         this.value = value;
         this.skipToTarget = skipToTarget;
+        this.endSurvey = Boolean.TRUE.equals(endSurvey) ? Boolean.TRUE : null;
+    }
+    public SurveyRule(Operator operator, Object value, String skipToTarget) {
+        this(operator, value, skipToTarget, null);
+    }
+    public SurveyRule(Operator operator, Object value) {
+        this(operator, value, null, Boolean.TRUE);
     }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java:32]: Overloaded constructors can get really messy really fast. I strongly recommend using the builder pattern instead.

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java [hunk 2]
```diff
@@ -15,75 +18,64 @@
...
+    // This is enforced in rules validation.
+    private final String skipToTarget;
+    private final Boolean endSurvey;
     
-    public SurveyRule() {
-    }
-    public SurveyRule(Operator operator, Object value, String skipToTarget) {
+    @JsonCreator
+    private SurveyRule(@JsonProperty("operator") Operator operator, @JsonProperty("value") Object value,
+            @JsonProperty("skipTo") String skipToTarget, @JsonProperty("endSurvey") Boolean endSurvey) {
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java]: Is it worth adding a test for JSON serialization/deserialization?

File: test/org/sagebionetworks/bridge/validators/SurveyValidatorTest.java [hunk 3]
```diff
@@ -380,6 +383,84 @@ public void willValidateRuleReferencesToNonQuestions() {
...
+    @Test
+    public void cannotSetEndSurveyToFalse() throws Exception {
+        String json = TestUtils.createJson("{'operator':'eq','value':'No',"+
+                "'skipTo':'theend','endSurvey':false}");
+        SurveyRule rule = BridgeObjectMapper.get().readValue(json, SurveyRule.class);
+        assertNull(rule.getEndSurvey());
+        assertEquals(Operator.EQ, rule.getOperator());
+        assertEquals("theend", rule.getSkipToTarget());
+        assertEquals("No", rule.getValue());
+    }
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/validators/SurveyValidatorTest.java]: Should this test be in SurveyRuleTest instead? The logic in question being tested is in SurveyRule, not SurveyValidator?

File: app/org/sagebionetworks/bridge/validators/SurveyValidator.java [hunk 4]
```diff
@@ -132,11 +132,24 @@ private void validateRules(Errors errors, List<SurveyElement> elements) {
...
+                        rejectField(errors, "rule", "cannot have a skipTo target and an endSurvey property");
+                    } 
+                    // But must have either a skipTo target or an endSurvey property
+                    else if (rule.getSkipToTarget() == null && rule.getEndSurvey() == null) {
+                        rejectField(errors, "rule", "must have a skipTo target or an endSurvey property");
+                    }
+                    // Validate the endSurvey value cannot be false
+                    else if (Boolean.FALSE.equals(rule.getEndSurvey())) {
+                        rejectField(errors, "rule", "cannot set endSurvey to false");
+                    }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/validators/SurveyValidator.java]: As implemented in SurveyRule, it's impossible for SurveyRule.getEndSurvey() to be false. Do we still need this check here? If we're worried that a future change might accidentally make getEndSurvey()==false to be possible, would it be better to just allow it in SurveyRule and always validate it here? This way, we have the logic in a single place and can be easily tested.

File: app/org/sagebionetworks/bridge/validators/SurveyValidator.java [hunk 5]
```diff
@@ -132,11 +132,24 @@ private void validateRules(Errors errors, List<SurveyElement> elements) {
...
+                    else if (rule.getSkipToTarget() == null && rule.getEndSurvey() == null) {
+                        rejectField(errors, "rule", "must have a skipTo target or an endSurvey property");
+                    }
+                    // Validate the endSurvey value cannot be false
+                    else if (Boolean.FALSE.equals(rule.getEndSurvey())) {
+                        rejectField(errors, "rule", "cannot set endSurvey to false");
+                    } 
+                    // Otherwise we can assume there's a skipToTarget, start checking that by looking for back references.
+                    else if (alreadySeenIdentifiers.contains(rule.getSkipToTarget())) {
                         rejectField(errors, "rule", "back references question %s", rule.getSkipToTarget());
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/validators/SurveyValidator.java:146]: It doesn't look like there's currently a test for back references. How difficult would it be to add this?

File: test/org/sagebionetworks/bridge/validators/SurveyValidatorTest.java [hunk 6]
```diff
@@ -380,6 +383,84 @@ public void willValidateRuleReferencesToNonQuestions() {
     }
     
     @Test
+    public void willVerifyRuleDoesntHaveSkipToAndEndSurvey() throws Exception {
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/validators/SurveyValidatorTest.java:389]: Is it worth also adding a test to make sure a valid rule with getEndSurvey() passes validation?

File: app/org/sagebionetworks/bridge/validators/SurveyValidator.java [hunk 7]
```diff
@@ -146,10 +159,13 @@ private void validateRules(Errors errors, List<SurveyElement> elements) {
...
-                        errors.pushNestedPath("elements["+i+"]");
-                        rejectField(errors, "rule", "has a skipTo identifier that doesn't exist: %s", rule.getSkipToTarget());
-                        errors.popNestedPath();
+                    // This validation only applies to skipTo target rules.
+                    if (rule.getSkipToTarget() != null) {
+                        if (!alreadySeenIdentifiers.contains(rule.getSkipToTarget())) {
+                            errors.pushNestedPath("elements["+i+"]");
+                            rejectField(errors, "rule", "has a skipTo identifier that doesn't exist: %s", rule.getSkipToTarget());
+                            errors.popNestedPath();
+                        }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/validators/SurveyValidator.java:164]: Similarly, how hard would it be to add a test for a non-existent skip-to target?

File: test/org/sagebionetworks/bridge/dynamodb/DynamoUploadDaoTest.java [hunk 8]
```diff
@@ -18,7 +17,6 @@
 import org.springframework.test.context.ContextConfiguration;
 import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
 
-import org.sagebionetworks.bridge.exceptions.ConcurrentModificationException;
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoUploadDaoTest.java:28]: Pretty sure this is a merge error on your end. Can you double check?

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java (hunk 1 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java:32]: Switched to builder

File: app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java (hunk 2 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/models/surveys/SurveyRule.java]: Added to SurveyRuleTest and to DynamoSurveyTest.

File: test/org/sagebionetworks/bridge/validators/SurveyValidatorTest.java (hunk 3 above)
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/validators/SurveyValidatorTest.java]: Moved to SurveyRuleTest

File: app/org/sagebionetworks/bridge/validators/SurveyValidator.java (hunk 4 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/validators/SurveyValidator.java]: Removed

File: app/org/sagebionetworks/bridge/validators/SurveyValidator.java (hunk 5 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/validators/SurveyValidator.java:146]: Added test.

File: test/org/sagebionetworks/bridge/validators/SurveyValidatorTest.java (hunk 6 above)
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/validators/SurveyValidatorTest.java:389]: Added

File: app/org/sagebionetworks/bridge/validators/SurveyValidator.java (hunk 7 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/validators/SurveyValidator.java:164]: Added

File: test/org/sagebionetworks/bridge/dynamodb/DynamoUploadDaoTest.java (hunk 8 above)
Comment (alxdarksage) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoUploadDaoTest.java:28]: In a more recent merge, some code was added back to this test that requires that CME. Next check in should remove this removal.

**Pull request merged**